feat: 멀티 스텝 Book Form 워크플로우와 재사용 가능한 폼 컴포넌트 구현

FranzCho opened the pull request:
## 폼 워크플로우 & 컴포넌트 구조

- **멀티 스텝 Book Form** (5단계 순차 입력) 추가  
  - 단계 이동 컴포넌트: `StepBar`, `StepButton`, `BookForm`  
  - 각 단계는 별도 컴포넌트(`step1`, `step2` …)로 구성  
  - 유효한 단계만 접근 가능하도록 네비게이션 로직 구현   

- **별점 평가 섹션(`StarRatingSection`)** 구현  
  - 0.5 단위 선택 및 실시간 점수 표시 지원 (Step 2 적용)

- **데이터 미리보기 및 모바일 레이아웃 추가**  
  - `BookSection`, `MobileForm`, `MobileSection`  

---

## 재사용 가능한 폼 컴포넌트

- **커스텀 입력 요소 제작**  
  - `Input`, `Textarea` (유효성 검사 + 글자 수 카운트)  
  - `RadioGroup`  
  - `DateRange` (시작일/종료일 동기화 및 유효성 검사)  
  - `SwitchCases` (범용 조건부 렌더링)  

- **react-hook-form 통합**  
  - 상태 관리 및 유효성 검증 적용

FranzCho committed:
레이아웃 단계 완료

FranzCho committed:
~ing

FranzCho committed:
도서정보 완료(유효성 X), StepBar & BookForm 수정

FranzCho committed:
현재 1단계 2단계 완료

FranzCho committed:
스텝수정

brightparagon commented on src/_components/bar/StepBar.tsx:
```diff
@@ -0,0 +1,49 @@
+
+    if (stepId < currentStep) return true;
+
+    return validateUpToStep(stepId - 1, formData);
```
여기에 - 1 을 해서 보내는 것이 한번에 이해하기는 어려운 구조인 것 같네요!

brightparagon commented on src/_components/bar/StepBar.tsx:
```diff
@@ -0,0 +1,49 @@
+
+  return (
+    <div className="flex gap-[20px] p-4">
+      {steps.map(({ id, label }) => (
+        <StepButton
+          key={id}
+          page={label}
+          isActive={currentStep === id}
+          disabled={!isStepAccessible(id)}
+          onClick={() => handleStepClick(id)}
+        />
+      ))}
```
스텝 구조를 맵으로 표현하기 보다는 아래처럼 나타내서 시인성을 높여보면 어떨까요?

```typescript
<SwitchCases
  value={step}
  cases={{
    step1: <Step1 />
    ...
  }}
/>
```

brightparagon replied:
음.. 이건 모든 스텝에 대해서 CTA 버튼을 보여주는거군요

brightparagon replied:
한 스텝마다는 이전/다음 버튼으로 구성해주셔야 합니다 ㅎㅎ

brightparagon commented on src/_components/common/DateRange.tsx:
```diff
@@ -0,0 +1,66 @@
+  startMinDate?: string;
+}
+
+export default function DateRange({
```
이 컴포넌트가 rhf에 의존성이 있다는 것을 드러내기 위해서 이름에 RHF prefix를 붙여야할 것 같아요

brightparagon commented on src/_components/common/SwitchCases.tsx:
```diff
@@ -0,0 +1,17 @@
+  defaultCase?: React.ReactElement;
+}
+
+function SwitchCases<T extends string | number>({
```
엇 이미 구현하셨네요 🤔

brightparagon commented on src/pages/index.tsx:
```diff
@@ -1,20 +1,51 @@
+      console.log('파싱된 데이터:', parsedData);
+      methods.reset(parsedData);
+    }
+  }, []);
```
const methods = useForm(); 과 함께 관련된 코드들은 하나의 코드로 묶어서 캡슐화해주면 어떨까요?

brightparagon commented on src/utils/stepValidation.ts:
```diff
@@ -0,0 +1,123 @@
+}
+
+// 날짜 검증
+const isValidDate = (dateString: string): boolean => {
```
validate 자체는 rhf를 쓰고있기 때문에 그 점을 활용해야 자연스러울 것 같습니다.
가령, zod나 yup 으로 모든 유효성 룰을 한군데에서 관리하고, 그걸 rhf 에서 사용하도록 하는게 나을 것 같아요.

FranzCho committed:
useBookForm 으로 캡슐화

FranzCho committed:
유효성 검사 UI 구현 및 Zod 스키마 적용

- Input, RadioGroup 컴포넌트에 에러 표시 UI 추가 (빨간 테두리, 에러 메시지)
- useBookForm에 첫 번째 에러 필드 자동 포커스 기능 구현
- Zod 스키마 기반 폼 검증으로 전환하여 유지보수성 향상
- Step4 인용구 관리 시스템 완성 (useFieldArray, 조건부 페이지 번호 검증)
- MobileForm 디자인 개선 및 TypeScript 타입 안정성 확보

brightparagon commented on src/hooks/useBookForm.tsx:
```diff
@@ -39,5 +44,21 @@ export function useBookForm(): {
   }, [watchAll, isClient]);
 
+  // 첫 번째 에러 필드로 자동 포커스
+  useEffect(() => {
+    const errors = methods.formState.errors;
+    const errorKeys = Object.keys(errors);
+
+    if (errorKeys.length > 0) {
+      const firstErrorField = errorKeys[0];
+      const element = document.querySelector(`[name="${firstErrorField}"]`) as HTMLElement;
+
+      if (element) {
+        element.focus();
+        element.scrollIntoView({ behavior: 'smooth', block: 'center' });
+      }
+    }
+  }, [methods.formState.errors]);
```
이건 이미 mode가 onChange여서 자동으로 될 것 같군요

brightparagon commented:
나머지 이견 없습니다!
